### ISSUE TITLE
unsafe manual memory management influenced by "print!" or providing an explicit type signature

### ISSUE THREAD

**Issue by bluescreen303** [rust-lang/rust#83275]
I tried this code:

```rust
use std::io::{self, SeekFrom, prelude::*};
use std::mem;
use std::ops::{Deref, DerefMut};

pub trait ReadSeek: Read + Seek {}
impl<T> ReadSeek for T where T: Read + Seek {}

pub trait StreamContext {
    fn current_offset(&mut self) -> Result<u64, io::Error>;
}

impl<S: Seek> StreamContext for S {
    fn current_offset(&mut self) -> Result<u64, io::Error> {
        self.seek(SeekFrom::Current(0))
    }
}

pub struct Droppy<T, F: FnOnce(&mut T)> {
    inner: T,
    dropfn: F,
}

impl<T, F: FnOnce(&mut T)> Drop for Droppy<T, F> {
    fn drop(&mut self) {
        let dropfn = mem::replace(&mut self.dropfn, unsafe {
            mem::MaybeUninit::uninit().assume_init()
        });
        let mut inner = mem::replace(&mut self.inner, unsafe {
            mem::MaybeUninit::uninit().assume_init()
        });
        // remove this print statement to get double free error
        print!("");
        dropfn(&mut inner);
    }
}

impl<T, F: FnOnce(&mut T)> Droppy<T, F> {
    pub fn new(inner: T, dropfn: F) -> Self {
        Droppy { inner, dropfn }
    }
    pub fn into_inner(mut self) -> T {
        let dropfn = mem::replace(&mut self.dropfn, unsafe {
            mem::MaybeUninit::uninit().assume_init()
        });
        let mut inner = mem::replace(&mut self.inner, unsafe {
            mem::MaybeUninit::uninit().assume_init()
        });
        mem::forget(self);
        dropfn(&mut inner);
        inner
    }

    pub fn map<U, G: FnOnce(&mut U), X: FnOnce(T) -> U>(mut self, f: X, dropfn: G) -> Droppy<U, G> {
        let _ = mem::replace(&mut self.dropfn, unsafe {
            mem::MaybeUninit::uninit().assume_init()
        });
        let inner = mem::replace(&mut self.inner, unsafe {
            mem::MaybeUninit::uninit().assume_init()
        });
        mem::forget(self);
        Droppy {
            inner: f(inner),
            dropfn,
        }
    }

    pub fn map_self<X: FnOnce(T) -> T>(mut self, f: X) -> Droppy<T, F> {
        let dropfn = mem::replace(&mut self.dropfn, unsafe {
            mem::MaybeUninit::uninit().assume_init()
        });
        let inner = mem::replace(&mut self.inner, unsafe {
            mem::MaybeUninit::uninit().assume_init()
        });
        mem::forget(self);
        Droppy {
            inner: f(inner),
            dropfn,
        }
    }
}

impl<T, F: FnOnce(&mut T)> Deref for Droppy<T, F> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl<T, F: FnOnce(&mut T)> DerefMut for Droppy<T, F> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

pub struct EofGuard<R: Read>(Droppy<R, fn(&mut R)>);

impl<R: Read> EofGuard<R> {
    pub fn new(inner: R) -> Self {
        EofGuard(Droppy::new(inner, Self::destruct))
    }
    fn destruct(inner: &mut R) {
        io::copy(inner, &mut io::sink())
            .expect("EofGuard::destruct to be able to read the stream to the end");
    }
    pub fn into_inner(self) -> R {
        self.0.into_inner()
    }
    pub fn map<S: Read, F: FnOnce(R) -> S>(self, f: F) -> EofGuard<S> {
        // without explicitly giving q's type, I get pointer errors
        let q: Droppy<S, fn(&mut S)> = self.0.map(f, EofGuard::destruct);
        EofGuard(q)
    }
}

impl<R: Read> Deref for EofGuard<R> {
    type Target = Droppy<R, fn(&mut R)>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<R: Read> DerefMut for EofGuard<R> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::BufReader;
    use std::io::Cursor;

    const SAMPLE_SIZE: usize = 64 * 1024;

    #[derive(Clone, Copy)]
    struct Num(u8);

    #[test]
    fn droppy1() -> Result<(), io::Error> {
        let mut x: u8 = 0;
        {
            Droppy::new(Num(15), |Num(q)| x = *q);
        }
        assert_eq!(x, 15);
        Ok(())
    }

    #[test]
    fn droppy2() -> Result<(), io::Error> {
        let mut x: u8 = 0;
        {
            let d = Droppy::new(Num(15), |Num(q)| x = *q);
            let Num(o) = d.into_inner();
            assert_eq!(o, 15);
        }
        {
            let d = Droppy::new(Num(16), |Num(q)| x = *q);
            let Num(o) = d.into_inner();
            assert_eq!(o, 16);
        }
        assert_eq!(x, 16);
        Ok(())
    }

    fn drop_helper(foo: &mut Num) {
        assert!(foo.0 > 10);
        println!("drop_helper: {}", foo.0);
    }

    #[test]
    fn droppy3() -> Result<(), io::Error> {
        {
            let d = Droppy::new(Num(15), drop_helper);
            let Num(o) = d.into_inner();
            assert_eq!(o, 15);
        }
        {
            let d = Droppy::new(Num(16), drop_helper);
            let Num(o) = d.into_inner();
            assert_eq!(o, 16);
        }
        Ok(())
    }

    #[test]
    fn droppy4() -> Result<(), io::Error> {
        let mut x: u8 = 0;
        {
            let d = Droppy::new(Num(15), drop_helper);
            let e = d.map(|Num(x)| x * 2, |q| x = *q);
            let qq = e.into_inner();
            assert_eq!(qq, 30);
        }
        assert_eq!(x, 30);
        Ok(())
    }

    #[test]
    fn droppy5() -> Result<(), io::Error> {
        let d = Droppy::new(Num(18), drop_helper);
        let e = d.map(|Num(x)| Num(x * 2), drop_helper);
        let Num(qq) = e.into_inner();
        assert_eq!(qq, 36);
        Ok(())
    }

    fn test_reader() -> impl ReadSeek {
        Cursor::new(vec![0; SAMPLE_SIZE])
    }

    #[test]
    fn eofguard1() -> Result<(), io::Error> {
        let mut t = test_reader();
        assert_eq!(t.current_offset()?, 0);

        let mut buf = [0; 128];
        t.read(&mut buf)?;
        assert_eq!(t.current_offset()?, 128);
        {
            let mut g = EofGuard::new(&mut t);
            assert_eq!(g.current_offset()?, 128);
        }
        assert_eq!(t.current_offset()?, SAMPLE_SIZE as u64);
        Ok(())
    }

    #[test]
    fn eofguard2() -> Result<(), io::Error> {
        let mut t = test_reader();
        assert_eq!(t.current_offset()?, 0);

        let mut buf = [0; 128];
        t.read(&mut buf)?;
        assert_eq!(t.current_offset()?, 128);
        {
            let mut g = EofGuard::new(t);
            assert_eq!(g.current_offset()?, 128);
            t = g.into_inner();
        }
        assert_eq!(t.current_offset()?, SAMPLE_SIZE as u64);
        Ok(())
    }

    #[test]
    fn eofguard3() -> Result<(), io::Error> {
        let mut t = test_reader();
        assert_eq!(t.current_offset()?, 0);

        let mut buf = [0; 128];
        t.read(&mut buf)?;
        assert_eq!(t.current_offset()?, 128);
        let mut g = EofGuard::new(t);
        assert_eq!(g.current_offset()?, 128);
        t = g.into_inner();
        assert_eq!(t.current_offset()?, SAMPLE_SIZE as u64);
        Ok(())
    }

    #[test]
    fn eofguard4() -> Result<(), io::Error> {
        let mut t = test_reader();
        assert_eq!(t.current_offset()?, 0);

        let mut buf = [0; 128];
        t.read(&mut buf)?;
        assert_eq!(t.current_offset()?, 128);
        let g = EofGuard::new(&mut t);
        {
            let mut h = g.map(BufReader::new);
            assert_eq!(h.current_offset()?, 128);
        }
        assert_eq!(t.current_offset()?, SAMPLE_SIZE as u64);
        Ok(())
    }
}

fn main() {
    println!("Hello, world!");
}
```

It's not a minimal example, but this way it's easier to understand goals/outcomes.
`Droppy` is supposed to be a smart pointer to make sure some deconstruction takes place when a value goes out of scope (like `Drop`) but still allows explicitly getting the value out (as normally `Drop` types cannot be deconstructed).

`EofGuard` is an example use for `Droppy`. It wraps a `Read`er and makes sure that no matter what, the stream will read to the end. This is sometimes useful when parsing things in a loop when next iteration of the loop expects the previous one to be consumed in full.

Now, this code uses some `unsafe` tricks to forcefully deconstruct values. This is my first playing around with `unsafe` so I was very cautious that *double free* and *invalid pointer* errors would probably be caused by me. However, I'm pretty confident I never leak or double free memory.

The example - as is - now works fine when running `cargo test`. But only after I had to make 2 adjustments that totally don't make sense to me how they would influence the result. But they do, consistently.

I would expect it to be OK to remove the *print* statement on line 32, but if I do and run `cargo test` I get
```
double free or corruption (out)
```

Also, I would expect to be able to inline lines 111 and 112 into
```rust
EofGuard(self.0.map(f, EofGuard::destruct))
```
but if I do and run `cargo test` I get
```
munmap_chunk(): invalid pointer
```

So I do not understand how these changes can cause this behaviour. And if my `unsafe` trickery is wrong, I don't understand how these changes would mask the problem. So I'm assuming these are bugs. 

### Meta

`rustc --version --verbose`:
```
rustc 1.50.0 (cb75ad5db 2021-02-10)
binary: rustc
commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
commit-date: 2021-02-10
host: x86_64-unknown-linux-gnu
release: 1.50.0
```

**Comment by bluescreen303** [comment]
only when needing/using that data later, which I make sure never happens.

Just closing the issue because it does some unsafe things doesn't make sense! 

Providing an explicit signature or adding an empty print statement should not *consistently* turn undefined behaviour in the wanted direction. So at the very least explain why the undefined behaviour would be influenced by these.

**Comment by jonas-schievink** [comment]
> only when needing/using that data later, which I make sure never happens.

No, this is always immediate undefined behavior, because it violates the type's validity invariant. You can see this for yourself when running the tests in [miri](https://github.com/rust-lang/miri).

> Providing an explicit signature or adding an empty print statement should not consistently turn undefined behaviour in the wanted direction

There are no guarantees once there is undefined behavior in the program. The compiler is free to produce any behavior, it may or may not be influenced by arbitrary other parts of the program.

If the issue persists after fixing these bugs, it would be helpful to get a minimal reproduction of the issue, ideally with unnecessary tests, types, and generic parameters removed.

**Comment by bluescreen303** [comment]
ok thanks I will dig a bit deeper.

**Comment by bluescreen303** [comment]
I managed to fix my code and the weird behaviour indeed went away.
Thanks for pointing out *miri*, it's really helpful when naively playing around with unsafe constructs.

As it turned out, I misunderstood some of the semantics around uninitialized memory and its relation to destructors, so miri forcing me to read the docs more carefully did the trick. Thanks for pointing that out :)
